model_filemanager/download_models.py

```diff
@@ -0,0 +1,173 @@
+            The subdirectory within the main models directory where the model 
+            should be saved (e.g., 'checkpoints', 'loras', etc.).
+        progress_callback (Callable[[str, DownloadStatus], Awaitable[Any]]): 
+            An asynchronous function to call with progress updates.
+
+    Returns:
+        DownloadModelResult: The result of the download operation.
+    """
+    if not validate_model_subdirectory(model_sub_directory):
+        return DownloadModelResult(
+            DownloadStatusType.ERROR, 
+            "Invalid model subdirectory", 
+            False
+        )
+
+    file_path, relative_path = create_model_path(model_name, model_sub_directory, models_dir)
+    existing_file = await check_file_exists(file_path, model_name, progress_callback, relative_path)
+    if existing_file:
+        return existing_file
+
+    try:
+        status = DownloadStatus(DownloadStatusType.PENDING, 0, f"Starting download of {model_name}")
+        await progress_callback(relative_path, status)
+
+        response = await model_download_request(model_url)
+        if response.status != 200:
+            error_message = f"Failed to download {model_name}. Status code: {response.status}"
+            logging.error(error_message)
+            status = DownloadStatus(DownloadStatusType.ERROR, 0, error_message)
+            await progress_callback(relative_path, status)
+            return DownloadModelResult(DownloadStatusType.ERROR, error_message, False)
+
+        return await track_download_progress(response, file_path, model_name, progress_callback, relative_path)
+
+    except Exception as e:
+        logging.error(f"Error in downloading model: {e}")
+        return await handle_download_error(e, model_name, progress_callback, relative_path)
+    
+
+def create_model_path(model_name: str, model_directory: str, models_base_dir: str) -> Tuple[str, str]:
+    full_model_dir = os.path.join(models_base_dir, model_directory)
+    os.makedirs(full_model_dir, exist_ok=True)
+    file_path = os.path.join(full_model_dir, model_name)
+    relative_path = '/'.join([model_directory, model_name])
+    return file_path, relative_path
+
+async def check_file_exists(file_path: str, model_name: str, progress_callback: Callable[[str, DownloadStatus], Awaitable[Any]], relative_path: str) -> Optional[DownloadModelResult]:
+    if os.path.exists(file_path):
+        status = DownloadStatus(DownloadStatusType.COMPLETED, 100, f"{model_name} already exists")
+        await progress_callback(relative_path, status)
+        return DownloadModelResult(DownloadStatusType.COMPLETED, f"{model_name} already exists", True)
+    return None
+
+
+async def track_download_progress(response: aiohttp.ClientResponse, file_path: str, model_name: str, progress_callback: Callable[[str, DownloadStatus], Awaitable[Any]], relative_path: str, interval: float = 1.0) -> DownloadModelResult:
+    try:
+        total_size = int(response.headers.get('Content-Length', 0))
```

Comment:
Can we add another endpoint for frontend to query content length. It would be nice to show user what is the file size before the actual download.

Reply:
Is there a reason you want a backend API for that vs just getting it directly on the frontend? Seems like an unnecessary API to have on the backend.

Reply:
Sorry about that. We can definitely do that in the frontend if we have the URL.